sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/test/TestTableProvider.java

```diff
@@ -213,6 +241,108 @@ public boolean supportsProjects() {
+
+      SerializableFunction<Integer, Boolean> comparison;
+      // TODO: add support for expressions like:
+      //  =(CAST($3):INTEGER NOT NULL, 200)
+      switch (node.getKind()) {
+        case LESS_THAN:
+          comparison = i -> i < 0;
+          break;
+        case GREATER_THAN:
+          comparison = i -> i > 0;
+          break;
+        case LESS_THAN_OR_EQUAL:
+          comparison = i -> i <= 0;
+          break;
+        case GREATER_THAN_OR_EQUAL:
+          comparison = i -> i >= 0;
+          break;
+        case EQUALS:
+          comparison = i -> i == 0;
+          break;
+        case NOT_EQUALS:
+          comparison = i -> i != 0;
+          break;
+        default:
+          throw new RuntimeException("Unsupported node kind: " + node.getKind().toString());
+      }
+
+      return Filter.<Row>create()
+          .whereFieldIds(
+              fieldIds, createFilter(operands, fieldIds, inputRefs, literals, comparison));
+    }
+
+    /**
+     * A helper method to create a serializable function comparing row fields.
+     *
+     * @param operands A list of operands used in a comparison.
+     * @param fieldIds A list of operand ids.
+     * @param inputRefs A list of operands, which are an instanceof {@code RexInputRef}.
+     * @param literals A list of operands, which are an instanceof {@code RexLiteral}.
+     * @param comparison A comparison to perform between operands.
+     * @return A filter comparing row fields to literals/other fields.
+     */
+    private SerializableFunction<Row, Boolean> createFilter(
+        List<RexNode> operands,
+        List<Integer> fieldIds,
+        List<RexInputRef> inputRefs,
+        List<RexLiteral> literals,
+        SerializableFunction<Integer, Boolean> comparison) {
+      if (inputRefs.size() == 2) { // Comparing 2 columns.
+        final int op0 = fieldIds.indexOf(inputRefs.get(0).getIndex());
+        final int op1 = fieldIds.indexOf(inputRefs.get(1).getIndex());
+        return row -> comparison.apply(row.<Comparable>getValue(op0).compareTo(op1));
+      } else { // Comparing a column to a literal.
+        int fieldSchemaIndex = inputRefs.get(0).getIndex();
+        FieldType beamFieldType = getSchema().getField(fieldSchemaIndex).getType();
+        final int op0 = fieldIds.indexOf(fieldSchemaIndex);
+
+        // Find Java type of the op0 in Schema
+        final Comparable op1 =
+            literals
+                .get(0)
+                .<Comparable>getValueAs(
+                    FieldTypeDescriptors.javaTypeForFieldType(beamFieldType).getRawType());
+        if (operands.get(0) instanceof RexLiteral) { // First operand is a literal
+          return row -> comparison.apply(op1.compareTo(row.getValue(op0)));
+        } else { // First operand is a column value
+          return row -> comparison.apply(row.<Comparable>getValue(op0).compareTo(op1));
+        }
+      }
+      // Case where we compare 2 Literals should never appear and get optimized away.
```

Comment:
That is true, but bugs happen. It is probably worth making the above `else` an `else if` and adding a `else` here that throws an exception.

Reply:
Added an exception when unexpected arguments are passed.

Reply:
Ok, I might misunderstand this comment. It appears to be about the `inputRefs.size() == 0` case? If that happens for some reason this will fail on `inputRefs.get(0)` above. What I think you should add is something like this:
```
assert inputRefs.size() == 1;
assert literals.size() == 1;
assert operands.size() == 1;
```
Seeing as `literals` and `operands` are always 0 or 1, you can simplify this by removing the List.

Reply:
Added a following list of asserts:
```
assert operands.size() == 2;
assert inputRefs.size() <= 2;
assert literals.size() < 2;
```
They should cover following scenarios:
1. Filter push-down for TestTable only supports comparisons between 2 operands.
2. Comparing two columns (2 input refs).
3. Comparing a column to a literal (1 input ref, 1 literal).